Add a ComputeReviewState function

mrnugget opened the pull request:
This adds a ComputeReviewState function that's to be used to compute the
review state.

It also brings back the ReviewState method on Changeset, except that
it's private now and only called for Bitbucket Server changesets.



<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->

ryanslade commented on internal/campaigns/types.go:
```diff
@@ -567,6 +567,37 @@ func (c *Changeset) Labels() []ChangesetLabel {
+// ReviewState of a Changeset. GitHub doesn't keep the review state on a
+// changeset, so it a GitHub Changeset will always return
+// ChangesetReviewStatePending.
+// This method should not be called directly. Instead call
```
Missing comment

ryanslade commented on internal/campaigns/types.go:
```diff
@@ -567,6 +567,37 @@ func (c *Changeset) Labels() []ChangesetLabel {
 }
 
+// ReviewState of a Changeset. GitHub doesn't keep the review state on a
+// changeset, so it a GitHub Changeset will always return
```
```suggestion
// changeset, so a GitHub Changeset will always return
```

ryanslade commented on enterprise/internal/campaigns/resolvers/resolver_test.go:
```diff
@@ -482,7 +482,7 @@ func TestCampaigns(t *testing.T) {
 					ServiceType: "bitbucketServer",
 				},
-				ReviewState: "APPROVED",
+				ReviewState: "PENDING",
```
Does this mean our old code was incorrect?

mrnugget replied:
Thanks for being so diligent! This was a nice catch. There _is_ a bug in the code. Currently investing.

mrnugget replied:
Investi**gating** btw! I'm not investing in bugs.

mrnugget commented:
@ryanslade Can you take another look? After you pointed it out, I dug into the tests and found out that we didn't handle Bitbucket's "Unapproved" event in the `(ChangesetEvents).ReviewState()` method. We _do_ handle it correct in the burndown chart. 
I added tests and fixed them by taking the event type into account.

I also added some mocks to the resolver tests so that they're less noisy.

ryanslade commented on internal/campaigns/types.go:
```diff
@@ -694,6 +726,28 @@ func ComputeChangesetState(c *Changeset, events ChangesetEvents) (ChangesetState
+
+	// For other codehosts we check whether the Changeset is newer or the
+	// events and use the newest entity to get the reviewstate.
+	newestEvent := events[len(events)-1]
```
Maybe we should add a comment above the function specifying that events should be sorted

mrnugget replied:
Like this?

https://github.com/sourcegraph/sourcegraph/pull/8978/files#diff-fa5e1bd8d3af52b49e4ee71fd894780eR730

ryanslade replied:
🤦‍♂
